File: Collections/HashSet.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Collections
{
    /// <summary>
    /// Set implementation using a hash table. 
    /// </summary>
    public class HashSet<Key> : ISet<Key>
    {
        private int n = 0;  // The number of keys in the table (cached for performance)
        private int m;  // The number of chains, i.e. table.Length (written this way for convenience)
        private Node[] table;

        private class Node
        {
            public readonly Key key;  // key must only be assigned once
            public Node next;

            public Node(Key key, Node next)
            {
                this.key = key;
                this.next = next;
            }
        }

        public HashSet(int capacity = 4)
        {
            if (capacity < 1)
            {
                throw new ArgumentOutOfRangeException();
            }
            table = new Node[capacity];
            m = capacity;
        }

        private void Resize(int size)
        {
            // We cannot simply clone the table as <code>Hash</code> function changes 
            var temp = new HashSet<Key>(size);
            foreach (var key in this)
            {
                temp.Add(key);
            }
            table = temp.table;
            m = temp.m;  // n.b. n should be unchanged
        }

        private int Hash(Key key) => (key.GetHashCode() & 0x7fffffff) % m;

        public void Add(Key key)
        {
            if (key == null)
            {
                throw new ArgumentNullException();
            }

            // Double the table size if the average length of a chain is >= 10
            if (n >= 10 * m) Resize(2 * m);

            var i = Hash(key);
            var current = table[i];

            while (current != null)
            {
                if (key.Equals(current.key))
                {
                    return;
                }
                current = current.next;
            }
            // Not found
            n++;
            table[i] = new Node(key, table[i]);
        }

        public bool Contains(Key key)
        {
            if (key == null)
            {
                throw new ArgumentNullException();
            }

            var i = Hash(key);
            var current = table[i];

            while (current != null)
            {
                if (key.Equals(current.key))
                {
                    return true;
                }
                current = current.next;
            }
            return false;
        }

        public IEnumerator<Key> GetEnumerator()
        {
            Node node;

            foreach (var chain in table)
            {
                node = chain;
                while (node != null)
                {
                    yield return node.key;
                    node = node.next;
                }
            }
        }

        public bool IsEmpty() => n == 0;

        public void Remove(Key key)
        {
            if (key == null)
            {
                throw new ArgumentNullException();
            }

            Node Remove(Node node)
            {
                if (node == null)
                    throw new KeyNotFoundException();
                if (key.Equals(node.key))
                {
                    n--;
                    return node.next;
                }
                node.next = Remove(node.next);
                return node;
            }

            var i = Hash(key);
            table[i] = Remove(table[i]);

            // Halve the table size if the average length of a chain is <= 2
            if (m > 4 && n <= 2 * m) Resize(m / 2);
        }

        public int Size() => n;

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}

File: src/main/java/in/bfdes/collections/HashSet.java
package in.bfdes.collections;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.function.Function;

public class HashSet<K> implements Set<K> {
    private int size = 0;  // the number of key-value pairs in the table
    private Node[] table;

    private class Node {
        public final K key;
        public Node next;

        public Node(K key, Node next) {
            this.key = key;
            this.next = next;
        }
    }

    public HashSet() {
        this(4);
    }

    public HashSet(int capacity) {
        if (capacity < 1)
            throw new IllegalArgumentException();
        table = (Node[]) new Object[capacity];
    }

    @Override
    public void add(K key) {
        if (key == null)
            throw new IllegalArgumentException();

        // Double the table size if the average length of a chain is >= 10
        if (size >= 10 * table.length) resize(2 * table.length);

        var i = hash(key);
        var node = table[i];

        while (node != null) {
            if (key.equals(node.key))
                return;
            node = node.next;
        }
        // Not found
        size++;
        table[i] = new Node(key, table[i]);
    }

    @Override
    public boolean contains(K key) {
        if (key == null)
            throw new IllegalArgumentException();

        var i = hash(key);
        var node = table[i];

        while (node != null) {
            if (key.equals(node.key))
                return true;
            node = node.next;
        }
        return false;
    }

    @Override
    public void remove(K key) {
        if (key == null)
            throw new IllegalArgumentException();

        var delete = new Function<Node, Node>() {
            @Override
            public Node apply(Node node) {
                if (node == null)
                    return null;  // Not found
                if (key.equals(node.key)) {
                    size--;
                    return node.next;
                }
                node.next = apply(node.next);
                return node;
            }
        };

        var i = hash(key);
        table[i] = delete.apply(table[i]);

        // Halve the table size if the average length of a chain is <= 2
        if (size > 4 && size <= 2 * table.length) resize(size / 2);
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public Iterator<K> iterator() {
        return new Iterator<>() {
            private int nextChain = 0;
            private Node nextNode;

            @Override
            public boolean hasNext() {
                return nextNode != null || nextChain < table.length - 1;
            }

            @Override
            public K next() {
                if (!hasNext())
                    throw new NoSuchElementException();
                if (nextNode == null)
                    nextNode = table[nextChain++];
                var key = nextNode.key;
                nextNode = nextNode.next;
                return key;
            }
        };
    }

    private void resize(int newCapacity) {
        var temp = new HashSet<K>(newCapacity);
        for (var key : this)
            temp.add(key);
        table = temp.table;
    }

    private int hash(K key) {
        return (key.hashCode() & 0x7fffffff) % table.length;
    }
}
